node/src/daser.rs

```diff
@@ -0,0 +1,294 @@
+use crate::store::{Store, StoreError};
+
+const DEFAULT_MAX_SAMPLES_NEEDED: usize = 16;
+
+type Result<T, E = DaserError> = std::result::Result<T, E>;
+
+/// Representation of all the errors that can occur when interacting with the [`Daser`].
+#[derive(Debug, thiserror::Error)]
+pub enum DaserError {
+    /// An error propagated from the [`P2p`] module.
+    #[error(transparent)]
+    P2p(#[from] P2pError),
+
+    /// An error propagated from the [`Store`] module.
+    #[error(transparent)]
+    Store(#[from] StoreError),
+}
+
+/// Component responsible for data availability sampling of blocks from the network.
+pub struct Daser {
+    cancellation_token: CancellationToken,
+}
+
+/// Arguments used to configure the [`Daser`].
+pub struct DaserArgs<S>
+where
+    S: Store,
+{
+    /// Handler for the peer to peer messaging.
+    pub p2p: Arc<P2p>,
+    /// Headers storage.
+    pub store: Arc<S>,
+}
+
+impl Daser {
+    /// Create and start the [`Daser`].
+    pub fn start<S>(args: DaserArgs<S>) -> Result<Self>
+    where
+        S: Store + 'static,
+    {
+        let cancellation_token = CancellationToken::new();
+        let mut worker = Worker::new(args, cancellation_token.child_token())?;
+
+        spawn(async move {
+            if let Err(e) = worker.run().await {
+                error!("Fatal DASer error: {e}");
+            }
+        });
+
+        Ok(Daser { cancellation_token })
+    }
+
+    /// Stop the [`Daser`].
+    pub fn stop(&self) {
+        // Singal the Worker to stop.
+        // TODO: Should we wait for the Worker to stop?
+        self.cancellation_token.cancel();
+    }
+}
+
+impl Drop for Daser {
+    fn drop(&mut self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+struct Worker<S>
+where
+    S: Store + 'static,
+{
+    cancellation_token: CancellationToken,
+    p2p: Arc<P2p>,
+    store: Arc<S>,
+    max_samples_needed: usize,
+}
+
+impl<S> Worker<S>
+where
+    S: Store,
+{
+    fn new(args: DaserArgs<S>, cancellation_token: CancellationToken) -> Result<Worker<S>> {
+        Ok(Worker {
+            cancellation_token,
+            p2p: args.p2p,
+            store: args.store,
+            max_samples_needed: DEFAULT_MAX_SAMPLES_NEEDED,
+        })
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        let cancellation_token = self.cancellation_token.clone();
+
+        loop {
+            select! {
+                _ = cancellation_token.cancelled() => break,
+                res = self.sample_next_block() => res?,
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn sample_next_block(&mut self) -> Result<()> {
+        let height = self.store.next_unsampled_height().await?;
+        self.store.wait_height(height).await?;
+
+        let header = self.store.get_by_height(height).await?;
+        let (sampled_cids, accepted) = self.sample_block(&header).await?;
+
+        self.store
+            .update_sampling_metadata(height, accepted, sampled_cids)
+            .await?;
```

Comment:
could you add a log here that we sampled some height and tell the result?

Reply:
done